daphne_worker/src/durable/mod.rs

```diff
@@ -209,97 +190,185 @@ impl<'srv> DurableConnector<'srv> {
-                    } else {
-                        // We only return an error if not in the "dev" deployment as
-                        // the experimental-local dev environment doesn't have
-                        // working get_alarm() and set_alarm() yet, so we want to
-                        // ignore errors in that case.
-                        return Err(e);
-                    }
-                }
+fn deployment(env: &Env) -> DaphneWorkerDeployment {
+    static DEPLOYMENT: OnceLock<DaphneWorkerDeployment> = OnceLock::new();
+
+    *DEPLOYMENT.get_or_init(|| {
+        let deployment = match env.var("DAP_DEPLOYMENT").map(|x| x.to_string()).as_deref() {
+            Ok("dev") => DaphneWorkerDeployment::Dev,
+            Ok("prod") | Err(_) => DaphneWorkerDeployment::Prod,
+            Ok(s) => {
+                tracing::error!("Invalid value for DAP_DEPLOYMENT ({s}), defaulting to prod");
+                DaphneWorkerDeployment::Prod
             }
-            *self.alarmed() = true;
+        };
+        if !matches!(deployment, DaphneWorkerDeployment::Prod) {
+            trace!("DAP deployment override applied: {deployment:?}");
         }
-        Ok(())
-    }
+        deployment
+    })
 }
 
-#[async_trait::async_trait(?Send)]
-trait GarbageCollectable: DapDurableObject {
-    /// A mutable property used to track whether this DO has been touched.
-    fn touched(&mut self) -> &mut bool;
-
-    fn env(&self) -> &Env;
+/// Generate a durable object based on a `DapDurableObject`.
+///
+/// This object must hold the `State` and `Env`. Thus the macro forces it.
+#[macro_export]
+macro_rules! mk_durable_object {
+    (struct $name:ident {
+        state: State,
+        env: Env,
+        $($field:ident : $type:ty),*
+        $(,)?
+    }) => {
+        #[worker::durable_object]
+        struct $name {
+            state: ::worker::State,
+            env: ::worker::Env,
+            $($field: $type),*
+        }
 
-    /// Run garbage collection requests.
-    ///
-    /// If a garbage collection request is handled no further processing needs to be done, as such,
-    /// this function will consume the request, otherwise it will return the passed in request for
-    /// further handling.
-    async fn schedule_for_garbage_collection(
-        &mut self,
-        req: Request,
-    ) -> Result<std::ops::ControlFlow<(), Request>> {
-        match GarbageCollector::try_from_uri(&req.path()) {
-            Some(GarbageCollector::DeleteAll) => {
-                self.state().storage().delete_all().await?;
-                *self.touched() = false;
-                return Ok(std::ops::ControlFlow::Break(()));
+        #[worker::durable_object]
+        impl DurableObject for $name {
+            fn new(state: ::worker::State, env: ::worker::Env) -> Self {
+                $crate::tracing_utils::initialize_tracing(&env);
+                <Self as $crate::durable::GcDurableObject>::new(state, env)
             }
```

Comment:
It's odd to shadow `new()` here. Can we use a different function name for the `GcDurableObject` constructor?